### PR TITLE
Storage refactoring and add better support for non-3d fields

### PR DESCRIPTION
**Pull request by egparedes**
## Description

Small storage refactoring to enhance support for non-3d fields (lower and higher dimensional fields).

### CODE REVIEW

File: tests/test_unittest/test_storage.py [hunk 1]
```diff
@@ -318,68 +318,92 @@ def test_allocate_gpu_unmanaged(param_dict):
...
+        normalize_storage_spec(default_origin, (10, 10), dtype, (False, False, False))
+
+    # Dtype
+    default_origin_out, shape_out, dtype_out, mask_out = normalize_storage_spec(
+        default_origin, shape, (dtype, (2,)), mask
+    )
+    assert default_origin_out == tuple([*default_origin, 0])
+    assert shape_out == tuple([*shape, 2])
+    assert dtype_out == dtype
+    assert mask_out == tuple([*mask, True])
```
Comment (gronerl) [inline tests/test_unittest/test_storage.py:406]: Github doesn't let me put this comment on the `test_masked_storage_?pu` below because it's not part of the commit, but there, the skip should now be removed again and the test fixed.

**Review by gronerl — CHANGES_REQUESTED** [review]
Looks mostly good, but the `test_masked_storage_?pu` should be turned back on now (and fixed), other comments are minor or a matter of taste.

File: src/gt4py/definitions.py [hunk 2]
```diff
@@ -57,7 +58,7 @@ def _check_value(cls, value, ndims):
         assert ndims[0] <= len(value) <= ndims[1]
 
     @classmethod
-    def is_valid(cls, value, *, ndims=(1, CartesianSpace.ndims)):
+    def is_valid(cls, value, *, ndims=(1, sys.maxsize)):
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/definitions.py:61]: Done

File: src/gt4py/storage/utils.py [hunk 3]
```diff
@@ -35,65 +36,72 @@ def idx_from_order(order):
...
+        # User-friendly axes specification (e.g. "IJK" or gtscript.IJK)
+        str_kind = "".join(str(i) for i in mask) if isinstance(mask, (tuple, list)) else str(mask)
+        axes_set = set(str_kind)
+        if axes_set - {"I", "J", "K"}:
+            raise ValueError(f"Invalid axes names in mask specification: '{mask}'")
+        if len(axes_set) != len(str_kind):
+            raise ValueError(f"Repeated axes names in mask specification: '{mask}'")
+        mask = ("I" in axes_set, "J" in axes_set, "K" in axes_set)
+    elif len(mask) < 3 and sum(mask):
+        mask = (*mask,) + (False,) * (3 - len(mask))
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/storage/utils.py:60]: Done

File: src/gt4py/storage/utils.py [hunk 4]
```diff
@@ -35,65 +36,72 @@ def idx_from_order(order):
...
+    elif not sum(mask):
+        raise ValueError(f"Invalid empty mask: '{mask}'")
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/storage/utils.py:62]: Done (although it probably does not make sense right now)

File: src/gt4py/storage/utils.py [hunk 5]
```diff
@@ -35,65 +36,72 @@ def idx_from_order(order):
...
+    if shape is not None:
+        if len(shape) < sum(mask):
+            raise ValueError(
+                f"len(shape)(={len(shape)}) cannot be lower than the number of 'True' entries in mask [{mask}]."
+            )
+        if not gt_util.is_iterable_of(shape, numbers.Integral):
+            raise TypeError("shape must be a tuple of ints or pairs of ints.")
+        if any(i <= 0 for i in shape):
+            raise ValueError(f"shape ({shape}) contains non-positive value.")
+        if sum(mask) < len(shape):
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/storage/utils.py:75]: Done

File: src/gt4py/storage/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/storage/utils.py:62]: Sure, but since it will not work out of the box right now, it's probably safe to forbid it.

File: tests/test_unittest/test_storage.py [hunk 6]
```diff
@@ -318,68 +318,92 @@ def test_allocate_gpu_unmanaged(param_dict):
...
+    default_origin_out, shape_out, dtype_out, mask_out = normalize_storage_spec(
+        default_origin, (10, 10), dtype, (True, True, False)
+    )
+    assert default_origin_out == (0, 0)
+    assert shape_out == (10, 10)
+    assert dtype_out == dtype
+    assert mask_out == (True, True, False)
+
+    _, shape_out, __, mask_out = normalize_storage_spec(
+        default_origin, (10, 20), dtype, (False, True, False)
```
[the remark this answers is not in the document]
Comment (egparedes) [reply tests/test_unittest/test_storage.py:366]: Done

**Comment by egparedes** [comment]
bors try

**Comment by gronerl** [comment]
bors try

**Review by gronerl — APPROVED** [review]

File: src/gt4py/utils/base.py [hunk 7]
```diff
@@ -217,7 +217,7 @@ def interpolate_mask(seq: Sequence[Any], mask: Sequence[bool], default) -> Tuple
     (0, 1, 0, 2)
     """
     it = iter(seq)
-    return tuple(it.__next__() if m else default for m in mask)
+    return tuple(next(it) if m else default for m in mask)
```
[the remark this answers is not in the document]
Comment (jdahm) [reply src/gt4py/utils/base.py:220]: That was my hack, I think ;)

**Comment by gronerl** [comment]
bors try
